lib/livebook/runtime/k8s.ex

```diff
@@ -0,0 +1,559 @@
+  # Usually, nodes need to be configured to use the same distribution
+  # protocol (`-proto_dist`). We configure the Kubernetes node to use IPv6
+  # distribution (`-proto_dist inet6_tcp`). However, depending whether
+  # the local node runs IPv4 or IPv6 distribution, we configure the
+  # kubectl proxy to bind to a IPv4 or IPv6 loopback respectively. The
+  # proxy always communicates with the Kubernetes Pod over IPv6, as
+  # is the case with all internal networking. Consequently, regardless
+  # of the protocol used by the local node, the remote node perceives
+  # it as IPv6.
+  #
+  # Sidenote, a node using IPv6 distribution may accept connections
+  # from a node using IPv4, depending on the `:kernel` application
+  # configuration `inet_dist_listen_options` -> `ipv6_v6only`, which
+  # has OS-specific value. However, we don't rely on this here.
+
+  defstruct [:config, :node, :req, :server_pid, :lv_pid, :pod_name]
+
+  @type config :: %{
+          context: String.t(),
+          namespace: String.t(),
+          home_pvc: String.t(),
+          docker_tag: String.t(),
+          pod_template: String.t()
+        }
+
+  @type t :: %__MODULE__{
+          node: node() | nil,
+          req: Req.Request.t(),
+          server_pid: pid() | nil,
+          lv_pid: pid(),
+          pod_name: String.t() | nil
+        }
+
+  use GenServer, restart: :temporary
+
+  require Logger
+
+  @doc """
+  Returns a new runtime instance.
+  """
+  @spec new(config :: map(), req :: Req.Request.t()) :: t()
+  def new(config, req) do
+    %__MODULE__{config: config, req: req, lv_pid: self()}
+  end
+
+  def __connect__(runtime) do
+    {:ok, pid} =
+      DynamicSupervisor.start_child(Livebook.RuntimeSupervisor, {__MODULE__, {runtime, self()}})
+
+    pid
+  end
+
+  @doc false
+  def start_link({runtime, caller}) do
+    GenServer.start_link(__MODULE__, {runtime, caller})
+  end
+
+  @impl true
+  def init({runtime, caller}) do
+    state = %{primary_ref: nil}
+    {:ok, state, {:continue, {:init, runtime, caller}}}
+  end
+
+  @impl true
+  def handle_continue({:init, runtime, caller}, state) do
+    config = runtime.config
+    %{namespace: namespace, context: context} = config
+    req = runtime.req
+
+    kubeconfig =
+      if System.get_env("KUBERNETES_SERVICE_HOST"),
+        do: nil,
+        else: System.get_env("KUBECONFIG") || Path.join(System.user_home!(), ".kube/config")
+
+    cluster_data = get_cluster_data(kubeconfig)
+
+    runtime_data =
+      %{
+        node_base: cluster_data.node_base,
+        cookie: Node.get_cookie(),
+        dist_port: cluster_data.remote_port
+      }
+      |> :erlang.term_to_binary()
+      |> Base.encode64()
+
+    with {:ok, pod_name} <-
+           with_log(caller, "create pod", fn ->
+             create_pod(req, config, runtime_data, cluster_data.remote_port)
+           end),
+         {:ok, pod_ip} <-
+           with_pod_events(caller, "Waiting for pod to get ready", req, namespace, pod_name, fn ->
```

Comment:
Lowercase for consistency and making it shorter:

```suggestion
           with_pod_events(caller, "waiting for pod", req, namespace, pod_name, fn ->
```